### PR TITLE
feat(WebWorker) Add channel support to MessageBus

### LINKED ISSUE #3661: Add channel support to MessageBus
MessageBus should support some form of channel abstraction. We could have channels for angular modules (renderer, compiler, xhr, router, etc...), and then users could specify their own channels.

### PR DESCRIPTION
**Pull request by jteplitz**
closes #3661 and #3686

### CODE REVIEW

File: modules/angular2/src/web-workers/shared/message_bus.ts [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
-export interface MessageBusSink { send(message: Object): void; }
+export interface MessageBusSink {
+  /**
+   * Returns an {@link EventEmitter} for the given channel
+   * To publish methods to that channel just call next (or add in dart) on the returned emitter
+   */
+  to(channel: string): EventEmitter;
+}
+
+export class MessageBus implements MessageBusInterface {
```
Comment (yjbanov) [inline modules/angular2/src/web-workers/shared/message_bus.ts]: leave TODO to remove this once we have abstract classes in TypeScript

File: modules/angular2/src/web-workers/shared/isolate_message_bus.dart [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  final Map<String, EventEmitter> _channels = new Map<String, EventEmitter>();
+
+  IsolateMessageBusSink(SendPort port) : _port = port;
+
+  EventEmitter to(String channel) {
+    if (_channels.containsKey(channel)) {
+      return _channels[channel];
+    } else {
+      var emitter = new EventEmitter();
+      ObservableWrapper.subscribe(emitter, (message) {
```
Comment (yjbanov) [inline modules/angular2/src/web-workers/shared/isolate_message_bus.dart]: No need to use facades in Dart code.

File: modules/angular2/src/web-workers/shared/isolate_message_bus.dart [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      : rawDataStream = port.asBroadcastStream() {
+    rawDataStream.listen((message) {
+      if (message is SendPort){
+        return;
+      }
+
+      if (message.containsKey("channel")) {
+        var channel = message['channel'];
+        if (_channels.containsKey(channel)) {
+          ObservableWrapper.callNext(_channels[channel], message['message']);
```
Comment (yjbanov) [inline modules/angular2/src/web-workers/shared/isolate_message_bus.dart]: ditto

File: modules/angular2/src/web-workers/ui/impl.ts [hunk 4]
```diff
@@ -360,3 +51,5 @@ class ReceivedMessage {
     this.type = data['type'];
   }
 }
+
+export interface UIComponent { attachToMessageBus(bus: MessageBus): void; }
```
Comment (yjbanov) [inline modules/angular2/src/web-workers/ui/impl.ts]: delete

**Comment by jteplitz602** [comment]
landed in 0b59e664ec0daa32ba6ed61af46b205fc3663d4b

**Pull request closed without merging**